presto-orc/src/main/java/com/facebook/presto/orc/zstd/FiniteStateEntropy.java

```diff
@@ -0,0 +1,168 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.zstd;
+
+import static com.facebook.presto.orc.zstd.BitStream.peekBits;
+import static com.facebook.presto.orc.zstd.FseTableReader.FSE_MAX_SYMBOL_VALUE;
+import static com.facebook.presto.orc.zstd.UnsafeUtil.UNSAFE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
+
+class FiniteStateEntropy
+{
+    private final FiniteStateEntropy.Table table;
+    private final FseTableReader reader = new FseTableReader();
+
+    public FiniteStateEntropy(int maxLog)
+    {
+        table = new FiniteStateEntropy.Table(maxLog);
+    }
+
+    public int decompress(final Object inputBase, final long inputAddress, final long inputLimit, byte[] weights)
+    {
+        long input = inputAddress;
+        input += reader.readFseTable(table, FSE_MAX_SYMBOL_VALUE, inputBase, input, inputLimit);
+
+        final Object outputBase = weights;
+        final long outputAddress = ARRAY_BYTE_BASE_OFFSET;
+        final long outputLimit = outputAddress + weights.length;
+
+        long output = outputAddress;
+
+        // initialize bit stream
+        BitStream.Initializer initializer = new BitStream.Initializer(inputBase, input, inputLimit);
+        initializer.initialize();
+        int bitsConsumed = initializer.getBitsConsumed();
+        long currentAddress = initializer.getCurrentAddress();
+        long bits = initializer.getBits();
+
+        // initialize first FSE stream
+        int state1 = (int) peekBits(bitsConsumed, bits, table.tableLog);
+        bitsConsumed += table.tableLog;
+
+        BitStream.Loader loader = new BitStream.Loader(inputBase, input, currentAddress, bits, bitsConsumed);
+        loader.load();
+        bits = loader.getBits();
+        bitsConsumed = loader.getBitsConsumed();
+        currentAddress = loader.getCurrentAddress();
+
+        // initialize second FSE stream
+        int state2 = (int) peekBits(bitsConsumed, bits, table.tableLog);
+        bitsConsumed += table.tableLog;
+
+        loader = new BitStream.Loader(inputBase, input, currentAddress, bits, bitsConsumed);
+        loader.load();
+        bits = loader.getBits();
+        bitsConsumed = loader.getBitsConsumed();
+        currentAddress = loader.getCurrentAddress();
+
+        byte[] symbols = table.symbol;
+        byte[] numbersOfBits = table.numberOfBits;
+        int[] newStates = table.newState;
+
+        // decode 4 symbols per loop
+        while (output < outputLimit) {
+            UNSAFE.putByte(outputBase, output, symbols[state1]);
+            int numberOfBits = numbersOfBits[state1];
```

Comment:
Oh, heh.. that was intellij's name assignment when I inlined a function that implemented each of these blocks. It doesn't even need to be separate variables, since they are just used temporarily.